include/circt/Dialect/Verif/VerifOps.td

```diff
@@ -260,4 +262,91 @@ def YieldOp : VerifOp<"yield", [
   ];
 }
 
+//===----------------------------------------------------------------------===//
+// Verification intent ops
+//===----------------------------------------------------------------------===//
+
+/// BMC -- short for bounded model check.
+def BMC : I32EnumAttrCase<"BMC", 0, "bmc">;
+/// KIP -- short for k-Inductive-Proof.
+def KInductiveProof : I32EnumAttrCase<"KInductiveProof", 1, "kinduction">;
+
+def FormalTestTypeAttr : I32EnumAttr<"FormalTestType", "formal test type",
+                                [BMC, KInductiveProof]> {
+  let cppNamespace = "circt::verif";
+}
+
+def FormalOp : VerifOp<"formal", [NoTerminator, RegionKindInterface]> {
+  let summary = "defines a formal verification test";
+  let description = [{
+    This operation defines a formal verification test with potentially additional
+    assertions and assumptions. This should be written by declaring symbolic 
+    values that are then connected to a hardware instance. These symbols can then
+    be reasoned about in additional assertions. The region will then be converted 
+    into btor2, SystemVerilog, or used for verification directly in circt-bmc. 
+    The operations in this regiion are ignored during regular SystemVerilog 
+    emission.   
+      
+    Additional information about how the test should be conducted can be passed 
+    using the two supported attributes. These attributes are `k`, which defines 
+    the bounds of the test, and `type` which defines the type of formal test 
+    being done, i.e. Bounded Model Checking (BMC) or k-Induction.
+  }];
+
+  let arguments = (ins OptionalAttr<APIntAttr>:$k,
+                       OptionalAttr<FormalTestTypeAttr>:$test_type,
+                       OptionalAttr<InnerSymAttr>:$inner_sym);
```

Comment:
I think this operation should have a non-optional symbol. Not sure if we should add a regular MLIR symbol or make the inner symbol mandatory, though. I'd just follow what other ops do.

Reply:
I see, so have a symbol be auto-generated if not given by the user? The goal is to basically allow for someone to expree 
```
formal(k=20, type=Formal.BMC) { ... }  
```
in Chisel without requiring it to be stored anywhere -- but I guess we need to be able to refer to it somehow in the error report so maybe you're right

Reply:
I'd make it mandatory for the user to provide a unique name for the test. Or is there a particular reason that shouldn't be the case?
E.g., in Rust (and other languages) a unit test is also a function that has a unique name with an attribute/annotation that it is a test.